### PR TITLE
[Merged by Bors] - feat(algebra/hom/centroid): Centre of the Centroid of a *-ring is a *-ring

### PR DESCRIPTION
**Pull request by mans0954**
This PR shows that the centre of the centroid of a *-ring is a *-ring.

Previously submitted to mathlib as https://github.com/leanprover-community/mathlib/pull/18096

---

- [x] depends on: #6562
- [x] depends on: #12924
- [x] depends on: #12938

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 1]
```diff
@@ -514,4 +515,43 @@ def commRing (h : ∀ a b : α, (∀ r : α, a * r * b = 0) → a = 0 ∨ b = 0)
...
+  star := fun f => {
+    toFun := fun a => star (f (star a))
+    map_zero' := by
+      simp only [star_zero, map_zero]
+    map_add' := fun a b => by simp only [star_add, map_add]
+    map_mul_left' := fun a b => by simp only [star_mul, map_mul_right, star_star]
+    map_mul_right' := fun a b => by simp only [star_mul, map_mul_left, star_star]
+  }
+
+}
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:534]: I think this is closed to the preferred style:
```suggestion
instance : Star (CentroidHom α) where
  star f :=
  { toFun := fun a => star (f (star a))
    map_zero' := by
      simp only [star_zero, map_zero]
    map_add' := fun a b => by simp only [star_add, map_add]
    map_mul_left' := fun a b => by simp only [star_mul, map_mul_right, star_star]
    map_mul_right' := fun a b => by simp only [star_mul, map_mul_left, star_star] }
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 2]
```diff
@@ -514,4 +515,43 @@ def commRing (h : ∀ a b : α, (∀ r : α, a * r * b = 0) → a = 0 ∨ b = 0)
...
+
+@[simp] lemma star_apply (f : CentroidHom α) (a : α) : (star f) a = star (f (star a)) := rfl
+
+/--
+Let α be a star ring with commutative centroid. Then the centroid is a star ring.
+-/
+def star_ring (mul_comm : ∀ f g : CentroidHom α, f * g = g * f) : StarRing (CentroidHom α) :=
+{ star_involutive := fun f => by
+    ext
+    rw [star_apply, star_apply, star_star, star_star]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:545]: This and `star_add` doesn't need commutativity; can you promote the instance above to `StarAddMonoid`?

Comment (mans0954) [reply Mathlib/Algebra/Hom/Centroid.lean:541]: Is your proposal that we change the definition of `CentroidHom` from
```
tructure CentroidHom (α : Type*) [AddMonoid α] [Mul α] extends α →+ α where...
```
to
```
structure CentroidHom (α : Type*) [AddMonoid α] [Mul α] extends α →+ α where...
```
?

Comment (eric-wieser) [reply Mathlib/Algebra/Hom/Centroid.lean:541]: Not at all, it's that you add a separate `StarAddMonoid` instance

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 3]
```diff
@@ -542,7 +542,8 @@ def star_ring (mul_comm : ∀ f g : CentroidHom α, f * g = g * f) : StarRing (C
   star_mul := fun f g => by
     rw [mul_comm]
     ext
-    rw [star_apply, mul_apply, mul_apply, star_apply, star_apply, star_star]
+    rw [star_apply, mul_apply, mul_apply, star_apply]
+    simp
```
Comment (mans0954) [inline Mathlib/Algebra/Hom/Centroid.lean:546]: In the lean3 version of this PR https://github.com/leanprover-community/mathlib/pull/18096/files#diff-b9e451cbb3a5b6efa050e64d3f4896b5d3ed3c51e9f034259709f8372aaaa525R342 I could put
```
rw [star_apply, mul_apply, mul_apply, star_apply, star_apply, star_star]
```
But doing this in Mathlib4 results in:
```
 Build: Mathlib/Algebra/Hom/Centroid.lean#L538
(deterministic) timeout at 'whnf', maximum number of heartbeats (200000) has been reached (use 'set_option maxHeartbeats <num>' to set the limit)

```

File: Mathlib/Algebra/Hom/Centroid.lean (hunk 2 above)
Comment (mans0954) [reply Mathlib/Algebra/Hom/Centroid.lean]: Ah, I see what you mean. Done.

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 4]
```diff
@@ -532,21 +532,28 @@ instance : Star (CentroidHom α) where
 
 @[simp] lemma star_apply (f : CentroidHom α) (a : α) : (star f) a = star (f (star a)) := rfl
 
+instance starAddMonoidOfCentroidHom : StarAddMonoid (CentroidHom α) where
+  star_involutive := fun f => by
+    ext
+    rw [star_apply, star_apply, star_star, star_star]
+  star_add := fun f g => by
+    ext
+    rw [star_apply, add_apply, star_add, add_apply, star_apply, star_apply]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:541]:
```suggestion
instance instStarAddMonoid : StarAddMonoid (CentroidHom α) where
  star_involutive f := by
    ext
    rw [star_apply, star_apply, star_star, star_star]
  star_add f g := by
    ext
    rw [star_apply, add_apply, star_add, add_apply, star_apply, star_apply]
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 5]
```diff
@@ -532,21 +532,28 @@ instance : Star (CentroidHom α) where
...
 Let α be a star ring with commutative centroid. Then the centroid is a star ring.
 -/
-def star_ring (mul_comm : ∀ f g : CentroidHom α, f * g = g * f) : StarRing (CentroidHom α) :=
-{ star_involutive := fun f => by
-    ext
-    rw [star_apply, star_apply, star_star, star_star]
+def starRingOfCommCentroidHom (mul_comm : ∀ f g : CentroidHom α, f * g = g * f) :
+    StarRing (CentroidHom α) :=
+{
+  starAddMonoidOfCentroidHom with
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:549]:
```suggestion
def starRingOfCommCentroidHom (mul_comm : ∀ f g : CentroidHom α, f * g = g * f) :
    StarRing (CentroidHom α) where
  __ := starAddMonoidOfCentroidHom
```
(and drop the final `}`)

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 6]
```diff
@@ -532,21 +532,28 @@ instance : Star (CentroidHom α) where
...
   star_mul := fun f g => by
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:550]:
```suggestion
  star_mul f g := by
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 7]
```diff
@@ -514,4 +515,46 @@ def commRing (h : ∀ a b : α, (∀ r : α, a * r * b = 0) → a = 0 ∨ b = 0)
...
 end NonUnitalRing
 
+
+section StarRing
+
+variable [NonUnitalNonAssocSemiring α] [StarRing α]
+
+instance : Star (CentroidHom α) where
+  star f :=
+  { toFun := fun a => star (f (star a))
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Centroid.lean:525]: Can you add a docstring that states that `star f a = star (f (star a))`, and gives a reference for this definition? I've seen a handful of cases where there's no obvious canonical star structure, which a reference would allay any concerns over.

Comment (mans0954) [reply Mathlib/Algebra/Hom/Centroid.lean:525]: @eric-wieser I've been thinking about this. I haven't been able to find a precise reference, but from my perspective, this is a consequence of a *-ring being a *-triple. The definition of a centroid homomorphism `T` on a *-triple requires the existence of another map `T^*`, and a mild condition can be placed on the *-triple guaranteeing uniqueness of `T^*`. In the case of a *-triple generated from a *-ring, the algebraic centroid is a *-subalgebra of the triple centroid. A further mild condition makes them coincide. The involution I've defined here coincides with the triple centroid one when the uniqueness condition is in force.

Perhaps a better approach than this PR would be for me to develop the theory outlined above in Mathlib and then treat this as a special case?

Comment (mans0954) [reply Mathlib/Algebra/Hom/Centroid.lean:527]: Furthermore, this definition makes the natural (iso|homo)morphism defined in #7843 into a *-(iso|homo)morphism.

Comment (eric-wieser) [reply Mathlib/Algebra/Hom/Centroid.lean]: Can you add that  (iso|homo)morphism result to this PR?

File: Mathlib/Algebra/Star/NonUnitalSubalgebra.lean [hunk 8]
```diff
@@ -82,16 +82,166 @@ theorem coeSubtype : (subtype s : s → A) = Subtype.val :=
...
+theorem coe_copy (S : NonUnitalStarSubsemiring R) (s : Set R) (hs : s = ↑S) :
+    (S.copy s hs : Set R) = s :=
+  rfl
+
+theorem copy_eq (S : NonUnitalStarSubsemiring R) (s : Set R) (hs : s = ↑S) : S.copy s hs = S :=
+  SetLike.coe_injective hs
+
+section Center
+
+variable (R)
```
Comment (mans0954) [inline Mathlib/Algebra/Star/NonUnitalSubalgebra.lean:162]: Why does this not seem to have any affect on the following lemma?

File: Mathlib/Algebra/Hom/Centroid.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Algebra/Hom/Centroid.lean]: Looks like this is resolved by now.

Comment (mans0954) [reply Mathlib/Algebra/Hom/Centroid.lean]: The PR has changed a lot in the last 18 months(!) but I think the equivalent line is: https://github.com/leanprover-community/mathlib4/pull/6595/files#diff-4c5d1ef43cba4fe158b5e0ad5afbba8f622bd9479cbb954544f3526f15ae999fR121

File: Mathlib/Algebra/Star/CentroidHom.lean [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  star_involutive _ := ext (fun _ => by
+    rw [star_apply, star_apply, star_star, star_star])
+
+instance instStarAddMonoid : StarAddMonoid (CentroidHom α) where
+  star_involutive f := ext (fun _ => by
+    rw [star_apply, star_apply, star_star, star_star])
+  star_add f g := ext fun _ => star_add _ _
+
+instance : Star (Subsemiring.center (CentroidHom α)) where
+  star f := ⟨star (f : CentroidHom α), Subsemiring.mem_center_iff.mpr (fun g => ext (fun a =>
```
Comment (eric-wieser) [inline Mathlib/Algebra/Star/CentroidHom.lean:53]: Does this follow more directly from `Set.star_mem_center`? I'm not at lean right now, but happy to push some cleanups myself.

I wonder if we should have an inherited star instance on all submonoids / subrings

Comment (eric-wieser) [reply Mathlib/Algebra/Star/CentroidHom.lean:53]: Sorry, I now see this comment is silly.

File: Mathlib/Algebra/Star/CentroidHom.lean [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+/-- The canonical *-homomorphism embedding the center of `CentroidHom α` into `CentroidHom α`. -/
+def centerStarEmbedding : Subsemiring.center (CentroidHom α) →⋆ₙ+* CentroidHom α where
+  toNonUnitalRingHom :=
+    (SubsemiringClass.subtype (Subsemiring.center (CentroidHom α))).toNonUnitalRingHom
+  map_star' f := by
+    simp only [RingHom.toMonoidHom_eq_coe, OneHom.toFun_eq_coe, MonoidHom.toOneHom_coe,
+      MonoidHom.coe_coe, SubsemiringClass.coe_subtype]
+    exact rfl
+
+theorem inr_star (z : NonUnitalStarSubsemiring.center α) :
```
Comment (eric-wieser) [inline Mathlib/Algebra/Star/CentroidHom.lean:80]: What does `inr` refer to here? I think this should probably be called `star_centerToCentroidCenter`, or stated in reverse as `centerToCentroidCenter_star`.

File: Mathlib/Algebra/Star/CentroidHom.lean [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  star_mul f g := by
+    ext a
+    calc
+      star (f * g) a = star (g * f) a := by rw [CommMonoid.mul_comm f g]
+      _ = star (g (f (star a))) := rfl
+      _ = star (g (star (star (f (star a))))) := by simp only [star_star]
+      _ = (star g * star f) a := rfl
+
+/-- The canonical *-homomorphism embedding the center of `CentroidHom α` into `CentroidHom α`. -/
+def centerStarEmbedding : Subsemiring.center (CentroidHom α) →⋆ₙ+* CentroidHom α where
```
Comment (eric-wieser) [inline Mathlib/Algebra/Star/CentroidHom.lean:72]: Can you add a `simp, norm_cast` lemma  that this is just a coercion? I tried `simps`, but it generates garbage.

File: Mathlib/Algebra/Star/CentroidHom.lean [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    rw [mul_comm.comm, star_apply, mul_apply, mul_apply, star_apply, star_apply, star_star])
+
+end NonUnitalNonAssocStarSemiring
+
+section NonAssocStarSemiring
+
+variable [NonAssocSemiring α] [StarRing α]
+
+/-- The canonical isomorphism from the center of a (non-associative) semiring onto its centroid. -/
+def starCenterIsoCentroid : StarSubsemiring.center α ≃⋆+* CentroidHom α where
```
Comment (eric-wieser) [inline Mathlib/Algebra/Star/CentroidHom.lean:124]: Could you add `simp` lemmas for this too? `simps` generates `_symm_apply_coe` correctly, but leaves a stray `toFun` in the other lemma.

(As a reminder, you can check what `simps` does with `@[simps?]`)

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

File: Mathlib/Algebra/Star/CentroidHom.lean (hunk 10 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/CentroidHom.lean]: I can't remember now! Renamed.

File: Mathlib/Algebra/Star/CentroidHom.lean (hunk 11 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/CentroidHom.lean:72]: I tried:
```
@[simp, norm_cast]
lemma centerStarEmbedding_apply (T : Subsemiring.center (CentroidHom α)) : 
    centerStarEmbedding T = T.val := rfl
```
But got:
```
norm_cast: badly shaped lemma, rhs can't start with coe
  ↑T
```

File: Mathlib/Algebra/Star/CentroidHom.lean (hunk 12 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/CentroidHom.lean:124]: Like this? https://github.com/leanprover-community/mathlib4/pull/6595/commits/6f262c709d8dc083bb4685a085d1ae7159021052

File: Mathlib/Algebra/Star/CentroidHom.lean [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    ⟨T 1, by refine ⟨?_, ?_, ?_, ?_⟩; all_goals simp [← map_mul_left, ← map_mul_right]⟩
+  left_inv z := Subtype.ext <| by simp only [MulHom.toFun_eq_coe,
+    NonUnitalRingHom.coe_toMulHom, NonUnitalStarRingHom.coe_toNonUnitalRingHom,
+    starCenterToCentroid_apply, mul_one]
+  right_inv T := CentroidHom.ext <| fun _ => by
+    simp [starCenterToCentroid_apply, ← map_mul_right]
+
+@[simp]
+lemma starCenterIsoCentroid_apply (a : ↥(NonUnitalStarSubsemiring.center α)) :
+    CentroidHom.starCenterIsoCentroid a = CentroidHom.starCenterToCentroid.toFun a := rfl
```
Comment (eric-wieser) [inline Mathlib/Algebra/Star/CentroidHom.lean:136]:
```suggestion
    CentroidHom.starCenterIsoCentroid a = CentroidHom.starCenterToCentroid a := rfl
```
or is the no coercion on `Centroid`?

Comment (eric-wieser) [reply Mathlib/Algebra/Star/CentroidHom.lean:136]: If what you had was correct, you can use `simps` after all here.

Comment (mans0954) [reply Mathlib/Algebra/Star/CentroidHom.lean]: Sorry - I forgot to modify the statement.

**Pull request closed without merging**